### PR TITLE
Use modified version of Geom to load CORS textures from thor

### PR DESCRIPTION
**Pull request by Rovanion**
The data is visible in the blend shader which it fucks up thuroughly.

Alternative titles: "CORSing through the galaxy", "The man who wouldn't stop lifting.", "Anomolous material 3 - Hey look it's the exact same thing as the last film" or "Silicon Valley - git gud son".

### CODE REVIEW

**Comment by hansfilipelo** [comment]
I get white screen and errors in console (Chrome) when loading page: 

```
Uncaught TypeError: Cannot read property 'call' of undefined
    at Function.weather_magic.textures.load_data.cljs$core$IFn$_invoke$arity$variadic (http://localhost:3449/js/compiled/out/weather_magic/textures.js:137:132)
    at weather_magic$textures$load_data (http://localhost:3449/js/compiled/out/weather_magic/textures.js:128:41)
    at Function.weather_magic.textures.load_data_into_atom_and_return_key_BANG_.cljs$core$IFn$_invoke$arity$variadic (http://localhost:3449/js/compiled/out/weather_magic/textures.js:181:47)
    at weather_magic$textures$load_data_into_atom_and_return_key_BANG_ (http://localhost:3449/js/compiled/out/weather_magic/textures.js:173:72)
    at http://localhost:3449/js/compiled/out/weather_magic/state.js:81:239
-----------------------------
Uncaught Error: No protocol method ISwap.-swap! defined for type undefined: 
    at Object.cljs$core$missing_protocol [as missing_protocol] (http://localhost:3449/js/compiled/out/cljs/core.js:299:9)
    at Function.cljs.core._swap_BANG_.cljs$core$IFn$_invoke$arity$5 (http://localhost:3449/js/compiled/out/cljs/core.js:3063:17)
    at Function.cljs.core.swap_BANG_.cljs$core$IFn$_invoke$arity$variadic (http://localhost:3449/js/compiled/out/cljs/core.js:15649:30)
    at cljs$core$swap_BANG_ (http://localhost:3449/js/compiled/out/cljs/core.js:15607:29)
    at weather_magic$transforms$update_model_coords (http://localhost:3449/js/compiled/out/weather_magic/transforms.js:52:29)
    at weather_magic$transforms$update_lat_lon (http://localhost:3449/js/compiled/out/weather_magic/transforms.js:58:46)
    at weather_magic$core$draw_frame_BANG_ (http://localhost:3449/js/compiled/out/weather_magic/core.js:79:41)
    at http://localhost:3449/js/compiled/out/weather_magic/core.js:102:37
    at thi$ng$geom$gl$webgl$animator$animate_STAR_ (http://localhost:3449/js/compiled/out/thi/ng/geom/gl/webgl/animator.js:90:23)
    at Function.thi.ng.geom.gl.webgl.animator.animate.cljs$core$IFn$_invoke$arity$2 (http://localhost:3449/js/compiled/out/thi/ng/geom/gl/webgl/animator.js:101:23)
    at thi$ng$geom$gl$webgl$animator$animate (http://localhost:3449/js/compiled/out/thi/ng/geom/gl/webgl/animator.js:72:46)
    at Function.thi.ng.geom.gl.webgl.animator.animate.cljs$core$IFn$_invoke$arity$1 (http://localhost:3449/js/compiled/out/thi/ng/geom/gl/webgl/animator.js:82:46)
    at thi$ng$geom$gl$webgl$animator$animate (http://localhost:3449/js/compiled/out/thi/ng/geom/gl/webgl/animator.js:68:46)
    at http://localhost:3449/js/compiled/out/weather_magic/core.js:101:68
```

**Comment by Rovanion** [comment]
Well, that doesn't happen on my machine.

**Comment by yotomten** [comment]
I get the same problems as hansfilip

**Comment by Rovanion** [comment]
Do you get the issue with the code on branch hämta-riktig-data2?

File: src/weather_magic/shaders.cljs [hunk 1]
```diff
@@ -18,21 +18,16 @@
...
+     gl_FragColor = vec4(ambientCol, 1.0) + diffuse * vec4(lightCol, 1.0) * lam;
    }")
 
 (def blend-fs
   "void main() {
-     vec4 texture = texture2D(base, vUV);
-     vec4 temperature;
+     float lam = lambert(surfaceNormal(vNormal, normalMat), normalize(lightDir));
+     vec4 mapDiffuse = texture2D(base, vUV);
+     float temp = texture2D(trump, vUV).r * 3.0;
```
Comment (yotomten) [inline src/weather_magic/shaders.cljs]: Texturen borde inte heta trump när trump inte längre används

File: src/weather_magic/textures.cljs [hunk 2]
```diff
@@ -12,14 +12,14 @@
...
+                         :src    path
+                         :filter [glc/linear glc/linear]
+                         :cors   ""})]
     {:texture texture :loaded loaded}))
 
 (defn load-texture-if-needed
-  [gl-ctx textures & paths]
+  [textures gl-ctx & paths]
   "Load a texture from the given path into the given WebGL context and
   a reference to it along with an inticator as to whether the texture
```
Comment (yotomten) [inline src/weather_magic/textures.cljs]: indicator, not inticator

File: src/weather_magic/textures.cljs [hunk 3]
```diff
@@ -39,6 +39,35 @@
             (when-not (contains? textures name)
               {(keyword name) (load-texture gl-ctx path)})))))
 
+(defn load-data
+  "Loads data from thor into a texture. Returs a map with {:key
+  str :map texture-map} where :key holds how to find the newly loaded
+  texture in texture-map."
+  [texture-map gl-ctx & {:keys [variable request-params] :or {variable "temperature"}}]
```
Comment (yotomten) [inline src/weather_magic/textures.cljs:46]: why variable and not dimension?

File: src/weather_magic/textures.cljs [hunk 4]
```diff
@@ -39,6 +39,35 @@
...
+        url (str "http://thor.hfelo.se/api/" variable
+                 (util/map->query-string request-map))
+        key (keyword (util/get-filename url))]
+    {:key key :map (load-texture-if-needed texture-map gl-ctx url)}))
+
+(defn load-data-into-atom-and-return-key!
+  "Load a texture if needed and mutate the given atom to contain
+  it. Return the key of the newly loaded texture."
+  [texture-map-atom gl-ctx & {:keys [variable request-params] :or {variable "temperature"}}]
+  (let [retval (load-data @texture-map-atom gl-ctx :variable variable :request-params request-params)]
```
Comment (yotomten) [inline src/weather_magic/textures.cljs]: retval or ret-val?

File: src/weather_magic/textures.cljs [hunk 5]
```diff
@@ -39,6 +39,35 @@
...
+    (swap! texture-map-atom merge (:map retval))
+    (:key retval)))
+
 (defn load-base-textures
   [gl-ctx]
-  (load-texture-if-needed gl-ctx {} "img/earth.jpg" "img/trump.png"))
+  (def gl-ctx gl-ctx)
```
Comment (yotomten) [inline src/weather_magic/textures.cljs]: bort

**Review by yotomten — CHANGES_REQUESTED** [review]
Good job

File: src/weather_magic/shaders.cljs (hunk 1 above)
Comment (Rovanion) [reply src/weather_magic/shaders.cljs]: Fixat.

File: src/weather_magic/textures.cljs (hunk 2 above)
Comment (Rovanion) [reply src/weather_magic/textures.cljs]: Done.

File: src/weather_magic/textures.cljs (hunk 3 above)
Comment (Rovanion) [reply src/weather_magic/textures.cljs:46]: It's called variable on the back end.

File: src/weather_magic/textures.cljs (hunk 4 above)
Comment (Rovanion) [reply src/weather_magic/textures.cljs]: ret-val.

File: src/weather_magic/textures.cljs (hunk 5 above)
Comment (Rovanion) [reply src/weather_magic/textures.cljs]: Done.

**Review by yotomten — APPROVED** [review]

**Pull request merged**